### ISSUE TITLE
Improving `ibm_is_network_acl_rule` priority handling

### ISSUE THREAD

**Issue by MadJlzz** [IBM-Cloud/terraform-provider-ibm#3509]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add a new way to set rule priority for `ibm_is_network_acl_rule` ([documentation here](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/is_network_acl_rule))

From experience, we are usually creating rules with priority at the top. To do that, we have currently to pass the `rule_id` to set up that correctly. I am proposing to change a bit the behaviour of the `before` attribute suchs that if the value of the field is:
- `top`: the rule will be created with the highest priority
- `bottom`: the rule will be create with the least priority (actually, it's what you call **omitting**, this is a proposition for consistency)
- `rule_id`: the new rule will be created before the given **rule_id**

### New or Affected Resource(s)

* ibm_is_network_acl_rule

### Potential Terraform Configuration

To set the rule to the top:
```hcl
resource "ibm_is_network_acl_rule" "allow_all_ingress_rule" {
  network_acl = ibm_is_vpc.vpc.default_network_acl
  name        = "allow-all-ingress"
  action      = "allow"
  source      = "0.0.0.0/0"
  destination = "0.0.0.0/0"
  direction   = "inbound"
  before      = "top"
}
```

Alternatively, to set it to the bottom:
```hcl
resource "ibm_is_network_acl_rule" "allow_all_ingress_rule" {
  network_acl = ibm_is_vpc.vpc.default_network_acl
  name        = "allow-all-ingress"
  action      = "allow"
  source      = "0.0.0.0/0"
  destination = "0.0.0.0/0"
  direction   = "inbound"
  before      = "bottom"
}
```

A more complex solution but more readable could be:
```hcl
resource "ibm_is_network_acl_rule" "allow_all_ingress_rule" {
  network_acl = ibm_is_vpc.vpc.default_network_acl
  name        = "allow-all-ingress"
  action      = "allow"
  source      = "0.0.0.0/0"
  destination = "0.0.0.0/0"
  direction   = "inbound"
  priority = "top"
}

resource "ibm_is_network_acl_rule" "allow_all_ingress_rule_2" {
  network_acl = ibm_is_vpc.vpc.default_network_acl
  name        = "allow-all-ingress-2"
  action      = "allow"
  source      = "0.0.0.0/0"
  destination = "0.0.0.0/0"
  direction   = "inbound"
  priority = "after"
  reference_rule_id = ibm_is_network_acl_rule.allow_all_ingress_rule.rule_id
}
```

WDYT?

**Comment by deepaksibm** [comment]
Hi @MadJlzz Currently the API only supports 'before' to order the rules. And from terraform perspective as all the rule resources will be created in parallel, it will not be possible to find the rule this rule should be placed before.
As for the priority "after" with reference_rule_id, it is similar to just providing 'before' a rule_id - with a single attribute it looks simple.

**Comment by en-ree** [comment]
Hi,
we have run into the same issue and I'd like to add some things here:
As of today you can already provide `null` as value for `before` to create the rule at the end. This doesn't mean though that it will stay there forever. It is just appended by the time of rule creation. 
So you could say that this may not be `bottom` but maybe `append` instead. Same would then apply for `top` which would more be `insert_as_first`. 

As for the design with the `before` field. This may be a perfectly valid technical perspective but this is not the way a human thinks. It should also be possible to set `after` before humans start reading a list from the top.

Anyway:
The current ordering mechanism is not usable in a loop. E.g. if you have a structure like this:

```
resource "ibm_is_network_acl_rule" "rules" {
  for_each = var.rules

  network_acl = ibm_is_vpc.vpc.default_network_acl
  name        = each.value.name
  action      = each.value.action
  source      = each.value.source
  destination = each.value.destination
  direction   = each.value.direction
  [...]
}
```

The whole construct with `before` or `after` won't work anymore because Terraform cannot know the keys of the next or previous element because of multithreading. And we cannot even put a depenency in it because we have an array.

IMO there are two possible ways to solve that:
1) Introducing a property named "order" (number) which is used to order the rules. I see some problems with duplications and missing numbers here but it would be a good solution IMHO.
2) The other way would be to introduce a new resource with only `acl_id`, `rule_id` and `before` or `after` as properties where you can set the order for the ACL rules AFTER their creation. Just like the attachment of an existing VSI volume to an exising VSI.

Hope this helps. Feel free to contact me!


Kind regards,
Enrico Reetz
Solution Developer | IBM Global Cloud Acceleration Team | IBM Cloud

**Comment by MadJlzz** [comment]
Thanks everyone for the nice insights. I do think that `rules` property of [ibm_is_network_acl](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/resources/is_network_acl) solves a lot of issues for the most common cases.

From what we did internally, we are creating very often ACLs in multiple terraform plans because we like to split them by **_business domain_**.

Also, the fact that creating a VPC automatically creates a default ACL and that `ibm_is_network_acl_rule` is not super user friendly (to me) makes us create automatically one new ACL for the VPC/Subnets.

Last but not least, configuring newly created ressources often requires external connectivity. Because it is easier to create ACLs like you mentioned we are usually allowing everything in the default ACL and blocking everything at the end of the process by assigning a new ACL to the VPC/Subnet

Sorry I went a bit off the initial topic but I hope that the motivation of my question is a bit clearer now. 

Thanks!

**Comment by Nashluffy** [comment]
Even setting a rule in the UI with priority "Set to bottom" doesn't work if a new rule is added afterwards. For us, we wanted a deny all rule to be the very last. 

The key part is ignoring changes to the before field for our "dynamic" rules, where order does not matter.

```hcl
resource "ibm_is_network_acl_rule" "deny_all" {
  name = "denyall"
  network_acl = ibm_is_network_acl.multizone_acl.id
  action = "deny"
  direction = "inbound"
  source      = "0.0.0.0/0"
  destination = "0.0.0.0/0"
}


resource "ibm_is_network_acl_rule" "rules" {
  for_each = {for rule in local.rules: rule.name => rule}

  lifecycle {
    ignore_changes = [ before ]
  }
```

**Comment by astha-jain** [comment]
This looks good to close. If there persist any other problems, please open a new issue.
